src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/LinkedEditingRange/LinkedEditingRangeEndpoint.cs

```diff
@@ -0,0 +1,133 @@
+        {
+            if (projectSnapshotManagerDispatcher is null)
+            {
+                throw new ArgumentNullException(nameof(projectSnapshotManagerDispatcher));
+            }
+
+            if (documentResolver is null)
+            {
+                throw new ArgumentNullException(nameof(documentResolver));
+            }
+
+            if (tagHelperFactsService is null)
+            {
+                throw new ArgumentNullException(nameof(tagHelperFactsService));
+            }
+
+            _projectSnapshotManagerDispatcher = projectSnapshotManagerDispatcher;
+            _documentResolver = documentResolver;
+            _tagHelperFactsService = tagHelperFactsService;
+        }
+
+        public RegistrationExtensionResult GetRegistration()
+        {
+            const string AssociatedServerCapability = "linkedEditingRangeProvider";
+            var registrationOptions = new LinkedEditingRangeRegistrationOptions();
+            return new RegistrationExtensionResult(AssociatedServerCapability, registrationOptions);
+        }
+
+        public async Task<LinkedEditingRanges> Handle(
+            LinkedEditingRangeParams request,
+            CancellationToken cancellationToken)
+        {
+            var document = await _projectSnapshotManagerDispatcher.RunOnDispatcherThreadAsync(() =>
+            {
+                _documentResolver.TryResolveDocument(
+                    request.TextDocument.Uri.GetAbsoluteOrUNCPath(), out var documentSnapshot);
+
+                return documentSnapshot;
+            }, cancellationToken);
+
+            if (document is null || cancellationToken.IsCancellationRequested)
+            {
+                return null;
+            }
+
+            var codeDocument = await document.GetGeneratedOutputAsync();
+            if (codeDocument.IsUnsupported())
+            {
+                return null;
+            }
+
+            var location = await GetSourceLocation(request, document).ConfigureAwait(false);
+            var ancestors = GetAncestors(codeDocument, location);
+
+            // We only care if the user is within a TagHelper with a valid start and end tag.
+            if (_tagHelperFactsService.TryGetNearestAncestorStartAndEndTags(
+                    ancestors, out var startTag, out var endTag) &&
+                startTag.Name is not null && endTag.Name is not null)
+            {
+                var startSpan = startTag.Name.GetLinePositionSpan(codeDocument.Source);
+                var endSpan = endTag.Name.GetLinePositionSpan(codeDocument.Source);
+                var ranges = new Range[2] { startSpan.ToRange(), endSpan.ToRange() };
+
+                return new LinkedEditingRanges
+                {
+                    Ranges = ranges,
+                    WordPattern = _wordPattern
+                };
+            }
+
+            return null;
+
+            static async Task<SourceLocation> GetSourceLocation(
```

Comment:
We do this in loads of locations probably a good fundamentals issue in the future would be to make a common extension method for this and update every location that does similar Position -> SourceLocation translations.

Reply:
Agreed - filed https://github.com/dotnet/aspnetcore/issues/35679 to track.